web/api/v1/api.go

```diff
@@ -658,6 +658,10 @@ func (api *API) labelNames(r *http.Request) apiFuncResult {
 		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
 	}
 
+	hints := &storage.LabelHints{
+		Limit: limit + 1, // One more than the limit to ensure that warnings work correctly.
```

Comment:
Actually if no limit is set, `limit` will be `math.MaxInt` and this will overflow.
That case should be handled separately.  
Maybe we can add a `limitParamToLimitHint` util or sth.

Reply:
If we can align the meaning of "no limit" between them, it'd be even better, we can make `parseLimitParam` consider 0 as the no limit value (as we do for the hint) + add some doc comments to state that, we'll need to turn the `if len(...) > limit` checks into  `if limit >0 && len(...) > limit` (we can have an inline util for that), but at least in this way there is no confusion and no conversion to be done and the "no limit" can return more than `9*10^18` :)